### PR TITLE
Make binary operators work between fixed-point and floating args

### PR DESCRIPTION
**Pull request by pmattione-nvidia**
Some of the binary operators in cuDF don't work between fixed_point and floating-point numbers after [this earlier PR](https://github.com/rapidsai/cudf/pull/15438) removed the ability to construct and implicitly cast fixed_point numbers from floating point numbers. This PR restores that functionality by detecting and performing the necessary explicit casts, and adds tests for the supported operators. 

Note that the `binary_op_has_common_type` code is modeled after `has_common_type` found in traits.hpp. 

This closes [issue 16090](https://github.com/rapidsai/cudf/issues/16090) 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by JayjeetAtGithub — APPROVED** [review]
Looks good to me. Thanks @pmattione-nvidia

File: cpp/src/binaryop/compiled/binary_ops.cuh [hunk 1]
```diff
@@ -159,6 +166,7 @@ struct ops2_wrapper {
       TypeRhs y   = rhs.element<TypeRhs>(is_rhs_scalar ? 0 : i);
       auto result = [&]() {
         if constexpr (std::is_same_v<BinaryOperator, ops::NullEquals> or
+                      std::is_same_v<BinaryOperator, ops::NullNotEquals> or
```
Comment (karthikeyann) [inline cpp/src/binaryop/compiled/binary_ops.cuh:169]: Good catch.

**Review by karthikeyann — APPROVED** [review]
Looks good to me 👍 Nice work.

**Pull request merged**